Gifski/util.swift

```diff
@@ -1373,3 +1373,89 @@ extension Result {
 		}
 	}
 }
+
+
+final class Once {
+	private var lock = os_unfair_lock()
+	private var hasRun = false
+	private var value: Any?
+
+	/**
+	Executes the given closure only once. (Thread-safe)
+
+	Returns the value that the called closure returns the first (and only) time it's called.
+
+	```
+	final class Foo {
+		private let once = Once()
+
+		func bar() {
+			once.run {
+				print("Called only once")
+			}
+		}
+	}
+
+	let foo = Foo()
+	foo.bar()
+	foo.bar()
+	```
+
+	```
+	func process(_ text: String) -> String {
+		return text
+	}
+
+	let a = once.run {
+		process("a")
+	}
+
+	let b = once.run {
+		process("b")
+	}
+
+	print(a, b)
+	//=> "a a"
+	```
+	*/
+	func run<T>(_ closure: () throws -> T) rethrows -> T {
+		os_unfair_lock_lock(&lock)
+		defer {
+			os_unfair_lock_unlock(&lock)
+		}
+
+		guard !hasRun else {
+			return value as! T
+		}
+
+		hasRun = true
+
+		let returnValue = try closure()
+		value = returnValue
+		return returnValue
+	}
+
+	// TODO: Make it support `rethrows` so that if the input `function` is `throws` then the wrapped function becomes throwing too.
```

Comment:
@boyvanamstel Any idea how to make this work? Not important, but would definitely be useful to make the `wrap` function work with as many functions as possible.

Reply:
@sindresorhus Don't think you can mark a closure as rethrows, but can you just add an overload?

```swift
func wrap<T, U>(_ function: @escaping ((T) throws -> U)) -> ((T) throws -> U) {
  return { parameter in
    try self.run {
      try function(parameter)
    }
  }
}
```

Reply:
Good idea! I didn't think of that. Still annoying that Swift requires so much duplication. It's now 4 overloads...

Reply:
Yeah. Haven't tried it yet but with swift 5 and dynamic callable you could maybe make something like:

let onceClosure = Once(closure)
onceClosure("Some Value")

Where the dynamic callable func is rethrowing..

Reply:
Hmm, I don’t want to loose type-safety though. I’ll play with it when Swift 5 final is out.